### PR TITLE
Add execution branch for socket activation to correct LISTEN_PID

### PR DESCRIPTION
**Pull request by alopukhov**
Proposal on https://github.com/rootless-containers/rootlesskit/issues/448
Fixing LISTEN_PID env variable for target process if it equals to PID of [rootlless:parent] process.

Tested with dockerd. Extra $$ script is not required anymore.
```Shell
systemd-socket-activate -E "XDG_RUNTIME_DIR=${XDG_RUNTIME_DIR}" -l 12312 \
./rootlesskit/rootlesskit --net=slirp4netns --slirp4netns-sandbox=true \
--slirp4netns-seccomp=true \
--disable-host-loopback \
--port-driver=slirp4netns \
--copy-up=/etc \
--copy-up=/run \
sh -c 'rm /run/docker ; exec dockerd -H fd://'
```

Can be demonstated.
```Shell 
> (timeout 10 systemd-socket-activate -l 12345 ./rootlesskit --pidns sh -c 'echo "\$\$ is $$; LISTEN_PID is $LISTEN_PID"' &); sleep 1; netcat localhost 12345
Listening on [::]:12345 as 3.
Communication attempt on fd 3.
Execing ./rootlesskit (./rootlesskit --pidns sh -c echo "\$\$ is $$; LISTEN_PID is $LISTEN_PID")
$$ is 11; LISTEN_PID is 11
```

Example output for v2.1.0:
```Shell
(timeout 10 systemd-socket-activate -l 12345 ./rootlesskit --pidns sh -c 'echo "\$\$ is $$; LISTEN_PID is $LISTEN_PID"' &); sleep 1; netcat localhost 12345
Listening on [::]:12345 as 3.
Communication attempt on fd 3.
Execing ./rootlesskit (./rootlesskit --pidns sh -c echo "\$\$ is $$; LISTEN_PID is $LISTEN_PID")
$$ is 11; LISTEN_PID is 3868868
```

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
cc @charliemirabile

File: cmd/rootlesskit/main.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ import (
 	"github.com/sirupsen/logrus"
 	"github.com/urfave/cli/v2"
 
+    "github.com/rootless-containers/rootlesskit/v2/pkg/activation"
```
Comment (AkihiroSuda) [inline cmd/rootlesskit/main.go:17]:
```suggestion
    "github.com/rootless-containers/rootlesskit/v2/pkg/systemd/activation"
```

File: cmd/rootlesskit/main.go [hunk 2]
```diff
@@ -29,17 +30,24 @@ import (
 	"github.com/rootless-containers/rootlesskit/v2/pkg/version"
 )
 
+
+const (
+		pipeFDEnvKey              = "_ROOTLESSKIT_PIPEFD_UNDOCUMENTED"
+		childUseActivationEnvKey  = "_ROOTLESSKIT_CHILD_USE_ACTIVATION_UNDOCUMENTED"
+		runActivationEnvKey       = "_ROOTLESSKIT_RUN_ACTIVATION_UNDOCUMENTED"
```
Comment (AkihiroSuda) [inline cmd/rootlesskit/main.go:37]: Probably these ones should be like `_ROOTLESSKIT_SYSTEMD_ACTIVATION_%s_UNDOCUMENTED`

File: cmd/rootlesskit/main.go
Comment (AkihiroSuda) [inline cmd/rootlesskit/main.go:1]: https://github.com/rootless-containers/rootlesskit/blob/master/hack/integration-systemd-socket.sh should be updated

**Comment by charliemirabile** [comment]
Looks good. Thanks for picking up where I left off. I was hesitant to introduce another whole process into the already somewhat complex way that rootlesskit launches the app as part of #429 especially for a first contribution and given my lack of experience with Go.

It is unfortunate that fork and exec are so tightly coupled in Go (though this of course helps with making the language suitable for cross platform development for other projects), since in C, it would be as easy as inserting a call to setenv in the code between fork and exec whereas here we need to introduce a whole other program as a shim and successfully plumb all the relevant info through it, but your implementation looks sane.

File: cmd/rootlesskit/main.go (hunk 2 above)
Comment (alopukhov) [reply cmd/rootlesskit/main.go]: changed to _ROOTLESSKIT_SYSTEMD_ACTIVATION_*_UNDOCUMENTED

File: cmd/rootlesskit/main.go
Comment (alopukhov) [reply cmd/rootlesskit/main.go:1]: * Replaced socat with uuidd (the first thing i've founde for ubuntu 22)
* Added to actions if that's ok. Required additional packages: `uuid-runtime` and `systemd`  The second one is for `systemd-socket-activate`.

uuidd requires LISTEN_PID env variable to be correct. Given wrong env it fails like this:
```Shell
> systemd-socket-activate -l /tmp/uuidd.sock sh -c 'LISTEN_PID=42 exec rootlesskit uuidd --no-pid --no-fork --socket-activation' &
[1] 14
> Listening on /tmp/uuidd.sock as 3.

> uuidd -d -t -n 1 -s /tmp/uuidd.sock
Communication attempt on fd 3.
Execing sh (sh -c LISTEN_PID=42 exec rootlesskit uuidd --no-pid --no-fork --socket-activation)
uuidd: no file descriptors received, check systemctl status uuidd.socket
[rootlesskit:child ] error: command [uuidd --no-pid --no-fork --socket-activation] exited: exit status 1
[rootlesskit:parent] error: child exited: exit status 1
uuidd: error calling uuidd daemon (read count): Connection reset by peer
[1]+  Exit 1                  systemd-socket-activate -l /tmp/uuidd.sock sh -c 'LISTEN_PID=42 exec rootlesskit uuidd --no-pid --no-fork --socket-activation'
```

File: cmd/rootlesskit/main.go [hunk 3]
```diff
@@ -29,17 +30,24 @@ import (
...
-		parentEUIDEnvKey = "ROOTLESSKIT_PARENT_EUID" // documented
-		parentEGIDEnvKey = "ROOTLESSKIT_PARENT_EGID" // documented
-	)
+	iAmActivationHelper := os.Getenv(runActivationHelperEnvKey) == "true"
 	iAmChild := os.Getenv(pipeFDEnvKey) != ""
 	id := "parent"
 	if iAmChild {
 		id = "child " // padded to len("parent")
+	} else if iAmActivationHelper {
+	    id = "activ."
```
Comment (AkihiroSuda) [inline cmd/rootlesskit/main.go:50]: What is this dot symbol?

File: hack/integration-systemd-socket-check-env.sh [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
```
Comment (AkihiroSuda) [inline hack/integration-systemd-socket-check-env.sh:1]: Every bash script should have `set -eu -o pipefail`?

File: cmd/rootlesskit/main.go [hunk 5]
```diff
@@ -29,17 +30,24 @@ import (
...
+)
+
 func main() {
-	const (
-		pipeFDEnvKey     = "_ROOTLESSKIT_PIPEFD_UNDOCUMENTED"
-		stateDirEnvKey   = "ROOTLESSKIT_STATE_DIR"   // documented
-		parentEUIDEnvKey = "ROOTLESSKIT_PARENT_EUID" // documented
-		parentEGIDEnvKey = "ROOTLESSKIT_PARENT_EGID" // documented
-	)
+	iAmActivationHelper := os.Getenv(runActivationHelperEnvKey) == "true"
```
Comment (AkihiroSuda) [inline cmd/rootlesskit/main.go:44]: This should use `strconv.ParseBool`?

File: cmd/rootlesskit/main.go (hunk 3 above)
Comment (alopukhov) [reply cmd/rootlesskit/main.go]: Replaced with `activation_helper` literal.

**Comment by AkihiroSuda** [comment]
Could you squash the commits?

File: pkg/systemd/activation/activation.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  }
+  panic("should not reach here")
+}
+
+func selectExecutable(path string) (string, error) {
+  if strings.Contains(path, "/") {
+    return path, nil
+  }
+  return exec.LookPath(path)
+}
```
Comment (AkihiroSuda) [inline pkg/systemd/activation/activation.go:36]: Extra newline

File: cmd/rootlesskit/main.go [hunk 7]
```diff
@@ -29,17 +31,27 @@ import (
...
 func main() {
-	const (
-		pipeFDEnvKey     = "_ROOTLESSKIT_PIPEFD_UNDOCUMENTED"
-		stateDirEnvKey   = "ROOTLESSKIT_STATE_DIR"   // documented
-		parentEUIDEnvKey = "ROOTLESSKIT_PARENT_EUID" // documented
-		parentEGIDEnvKey = "ROOTLESSKIT_PARENT_EGID" // documented
-	)
+	iAmActivationHelper, err := strconv.ParseBool(os.Getenv(runActivationHelperEnvKey))
+	if err != nil {
+	  iAmActivationHelper = false
```
Comment (AkihiroSuda) [inline cmd/rootlesskit/main.go:47]: Should panic?

File: pkg/systemd/activation/activation.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    return err
+  }
+  panic("should not reach here")
+}
+
+func selectExecutable(path string) (string, error) {
+  if strings.Contains(path, "/") {
+    return path, nil
+  }
+  return exec.LookPath(path)
```
Comment (AkihiroSuda) [inline pkg/systemd/activation/activation.go:35]: Can't we unconditionally use `exec.LookPath`?

Comment (alopukhov) [reply pkg/systemd/activation/activation.go]: You are right. `if` branch is unnecessary.

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**